Unity-Technologies/com.unity.mobile.logcat
# Expand Screen Capture with the ability to record video

### Purpose of PR
**Type of change:**

* **Description of the feature**
I had this code for a while with the ability to capture video from android device, using process called screenrecord.
I think it's a very useful thing to have thus will try to push it to logcat package.

In short, I reworked Screen Capture window:
* It has now the ability to switch capture between screenshot and video
* Added Open button, so you can quickly open screenshot or video with associated app.


* **Links to screenshots, design docs, user docs, etc.**
![capture_screenshot](https://user-images.githubusercontent.com/568752/158167026-60540061-886f-4bd6-be25-fd36ae484e34.png)
![capture_video](https://user-images.githubusercontent.com/568752/158167039-e338a156-27e7-479f-9f87-0e1601512a1d.png)

* **Links to Jira/Fogbugz cases**
* **Halo effect**
Screen Capture window reworked
* **Performance Impact**
No performance impact
* **Package requirements**
Nothing changed in package requirements

### Checklist for PR maker
- [ ] Have you added a backport label? (if needed)
- [X ] Have you updated the Changelog? _Each package has a `CHANGELOG.md` file_
- [X] **Have you added or updated the Documentation to your PR?**
You can see new docs here https://github.com/Unity-Technologies/com.unity.mobile.logcat/blob/capture-video/com.unity.mobile.android-logcat/Documentation~/ScreenCapture.md

---
### Testing status
* New integration tests:
    * CanGetScreenshot
    * CanGetVideo
    * CanGetVideoWithTimeLimit
    * CaptureVideoHandlesErrors

The tests can be seen here - https://yamato.cds.internal.unity3d.com/jobs/277-com.unity.mobile.android-logcat/tree/capture-video/.yamato%252Fupm-integration-ci.yml%2523test_integration_2020.3/12259422/job

* **Detailed description of what testing was done, what projects were run**
First of all I'll copy paste some known issues from docs:
    * **screenrecord** doesn't capture sound, only image is captured
    * Logcat package uses Unity's video player for the video preview, you might see **WindowsVideoMedia** warnings in Editor console if your host OS is Windows, since displayed video is not processed
    * If you're capturing a static image, the video will contain only one frame and be a zero length video
    * **Chrome OS**: you can only capture the video from the sandboxed app, meaning which you install manually via adb or Unity. You cannot capture video from desktop or builtin apps.


* I tested on following devices:
    * Google Pixel 2 XL with Android 11
    * Chrome OS with Android 9 (see documentation for known issues)
    * Cube U83 with Android 6 (unlike Android 11, if screenrecord fails, exit code of the process is 0, which is unexpected, capture video accounts for that, but this case proves that different Android versions matter)

* Scenario 1:
    * In Screen Capture window, switch to Video
    * Click Capture
    * Wait 5 seconds
    * Click Stop
    * The video should be downloaded from the device and should appear in screen capture window running via Unity's video player

* Scenario 2:
    * In Screen Capture window, switch to Video
    * Click Time Limit toggle, and set it to 5 seconds
    * Click Capture
    * Wait 5 seconds, the video recording should stop automatically ! No need to hit Stop
    * The video should be downloaded from the device and should appear in screen capture window running via Unity's video player

* Scenario 3:
    * In Screen Capture window, switch to Video
    * Click Bit Rate, enter invalid value 0 
    * Click Capture
    * Recording should fail and proper error message should be displayed, proving that screen capture window displays errors correctly

* Scenario 4:
    * In Screen Capture window, switch to Video
    * Click Capture immediately followed by Stop
    * In some cases, Android will produce invalid video. If video preview tries to play such video, some errors a present in Editor console, which is kind a expected, since video is invalid. 


* **What environment did you test on?**
I tested on Windows and Android devices mentioned above.
* **Projects**
Used SampleProject1 from repo
* **Testing checklist**
- [ ] Built and run editor Locally or Katana
- [ ] Built a player Android/iOS (if applicable)
- [ ] Run on device Android/iOS (if applicable)
- [ ] If new feature has UI, does it match Unity style? (Unity [HIG](https://unitytech.github.io/unityeditor-hig/)?)
- [ ] Tested Undo/Redo + Prefab overrides + Alignment in Preset
- [X] All items have tooltips?
---

### Comments to reviewers
QA, Video Preview depends on host machine, for ex., on Windows it uses WindowsVideoMedia coded for playing video. That means on OSX it uses a different codec, could you check OSX please
